Ability to run plugin even on failure

twaugh opened the pull request:
Now plugins can determine whether the build has already been marked as failed.

This will be needed when we have BuildConfigs to prune.

TomasTomecek commented:
Shouldn't there be a code in `inner.py` which sets the variable appropriately (once the build has finished)? Also, wouldn't it make more sense to name it `build_has_failed`?

Maybe I just didn't get the patch.

twaugh commented:
Yes, `build_has_failed` probably makes more sense.

The point of this change is to allow a post-build plugin to decide how to behave based on whether this build has failed. That's why we have to set the variable as soon as a plugin fails that is not allowed to fail (https://github.com/projectatomic/atomic-reactor/commit/701bf3abdd20a8b11811697a6d53626c7f0ed7cf#diff-6efbaa8f7bea401e03542bd227209afbR218).

Setting it after all the plugins have run would mean no plugin gets to use it.

For an example of this in action, see https://github.com/projectatomic/atomic-reactor/pull/230.

TomasTomecek commented:
I'm sorry I still don't get it, let me show you how would I implement it.

TomasTomecek commented:
This is what I would expect this PR to contain: https://github.com/TomasTomecek/atomic-reactor/commit/990fcfb0c9144fee0dbe1e216b22e77518d78ff1

twaugh commented:
OK, I see what you mean now. Yes, it needs to do that.

However, looking more at that code I see we also need to have a type of plugin that *always* runs, even if (for example) a pre-build plugin failed, or the Docker build itself failed. Maybe it could be called `CleanupPlugin`.

I'll look at adding this.

TomasTomecek commented:
https://github.com/projectatomic/atomic-reactor/issues/110

twaugh commented:
Oh, didn't see that.

TomasTomecek commented on atomic_reactor/plugin.py:
```diff
@@ -210,6 +214,9 @@ def __init__(self, dt, workflow, plugin_class_name, plugins_conf, *args, **kwarg
         super(BuildPluginsRunner, self).__init__(plugin_class_name, plugins_conf, *args, **kwargs)
 
+    def on_plugin_failed(self):
+        self.workflow.build_has_failed = True
```
what does this mean?

twaugh replied:
It means that `build_inside()` will raise an exception.

TomasTomecek replied:
I'm sorry I still don't follow. I thought that setting `build_has_failed` makes only sense in `inner.py` where we figure out if the build successful.

twaugh replied:
This is about whether atomic-reactor succeeded or failed, not just whether the docker build did.

Some plugins can cause the entire run to be a failure, and there needs to be a hook to allow different behaviour in that case.

twaugh replied:
Let me explain another way. When we set up automated rebuilds, we only want to rebuild from Builds (OpenShift Build objects) that built successfully in the first place.

That's what this instance variable is intended to be: will this Build get marked by OpenShift as completed or failed?

TomasTomecek replied:
Right. Now I finally get it.

So the variable will reflect two states:
1. `docker build` has failed
2. one of the plugins has failed

Wouldn't it make sense rather use two variables for that?

twaugh replied:
From the point of view of an exit plugin wanting to know whether everything has gone to plan so far, though, it would be nice for it to have a single thing to check.

TomasTomecek commented:
Once we merge this, we should then transform these plugins to be exit plugins:
 * `remove_built_image` (removing base image)
 * `store_metadata_in_osv3`
 * `store_logs_to_file`

twaugh commented:
We need to decide which approach is best: exit plugins, or `run_always` plugin arguments.

twaugh commented:
> Once we merge this, we should then transform these plugins to be exit plugins:

Oh, we'll need to give exit plugins access to the workflow instance then.

twaugh commented:
I've changed the ExitPlugin argument from a `bool` to a `DockerBuildWorkflow` instance and squashed some commits to make the history nicer. I also changed those plugins to be exit plugins.

twaugh commented:
Oh, remove_built_image needs the tasker as well.

twaugh commented:
I've added a test case to show how exit plugins work. It also has machinery for checking the case where the first of two prebuild plugins fails, but it's commented out (last line). Currently the second plugin will still run... is that what we want?

twaugh commented:
(coveralls failure is due to pulp test no longer getting run, see https://github.com/release-engineering/dockpulp/pull/16)

twaugh commented:
I've changed the plugin failure hook so that it sets `plugin_failed` instead of `build_failed`, and added a `build_process_failed` property. CI failures are my fault but elsewhere :-(

TomasTomecek commented:
Very nice! Ready for merge?

twaugh commented:
Actually, no, I think the conversion of some plugins to exit plugins isn't complete. I'll fix that up.

twaugh commented:
Previously, an older osbs-client would try running what are now exit plugins as postbuild plugins, and they would not run. Now ExitPlugin inherits from PostBuildPlugin, and the plugins runner logs an error if an exit plugin is being run as a post-build plugin.

The upshot is that we now provide more compatibility for older osbs-client versions. Note: we still need to update osbs-client to actually run the exit plugins as exit plugins not post-build plugins.

This is ready for merge, with one open question: is this test valid? https://github.com/projectatomic/atomic-reactor/commit/0f218ecb8b3d2d98880dc1085406326a3e09a5d2#diff-1d4c7d01e22fd1091edb1a829562521cR261

twaugh commented:
Let's leave the issue of running all plugins of a certain type even when one fails until after this PR is merged. For now, I've added a test to verify the current behaviour.

Ready for merge.

twaugh commented:
Fixes #110.

TomasTomecek commented:
I thought this is already merged. Can you rebase please? (I should have merged this first, sorry about that)

twaugh commented:
Rebased and fixed a conflict caused by the movement of `test_rpmqa_plugin()`.
